python/tchannel/tornado/dispatch.py

```diff
@@ -239,50 +229,138 @@ def __init__(
+    def get_body_s(self):
+        """Get the raw stream of body.
 
-        :return: return the future object contains the value for arg1
+        :return: the argstream of body
         """
-        return get_arg(self, 0)
+        return self.argstreams[2]
 
-    def arg2(self):
-        """get value for arg2
+    def get_header(self):
+        """Get the header value from the request.
 
-        :return: return the future object contains the value for arg2
+        :return: the value of header
         """
         return get_arg(self, 1)
 
-    def arg3(self):
-        """get value for arg3
+    def get_body(self):
+        """Get the body value from the request.
 
-        :return: return the future object contains the value for arg3
+        :return: the value of body
         """
         return get_arg(self, 2)
 
-    def args(self):
-        """get value for arg1, arg2, and arg3
+    def set_body(self, stream):
+        """Set customized body stream.
 
-        :return: return the future object contains the list
-        for arg1, arg2, arg3
+        Note: the body stream can only be changed before the stream
+        is consumed.
+
+        :param stream: InMemStream/PipeStream for body
+
+        :except TChannelException:
+            Raise TChannelException if the stream is being sent when you try
+            to change the stream.
+        """
+        if self.argstreams[2].state == StreamState.init:
+            self.argstreams[2] = stream
+        else:
+            raise TChannelException(
+                "Unable to change the body since the streaming has started")
+
+    def set_header(self, stream):
+        """Set customized header stream.
+
+        Note: the header stream can only be changed before the stream
+        is consumed.
+
+        :param stream: InMemStream/PipeStream for header
+
+        :except TChannelException:
+            Raise TChannelException if the stream is being sent when you try
+            to change the stream.
         """
-        return [self.arg1(), self.arg2(), self.arg3()]
+
+        if self.argstreams[1].state == StreamState.init:
+            self.argstreams[1] = stream
+        else:
+            raise TChannelException(
+                "Unable to change the header since the streaming has started")
+
+    @tornado.gen.coroutine
+    def write_header(self, chunk):
+        """Write to header.
+
+        Note: the header stream is only available to write before write body.
+
+        :param chunk: content to write to header
+
+        :except TChannelException:
+            Raise TChannelException if the response's flush() has been called
+        """
+
+        if self.flushed:
+            raise TChannelException("write operation invalid after flush call")
+
+        if (self.argstreams[0].state != StreamState.completed and
+                self.argstreams[0].auto_close):
+            self.argstreams[0].close()
+
+        yield self.argstreams[1].write(chunk)
```

Comment:
Don't need to make these functions coroutines, you can just return the future here instead of yielding to it.